moveit_core/collision_detection_bullet/src/collision_robot_bt.cpp

```diff
@@ -0,0 +1,251 @@
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Author: Ioan Sucan */
+
+#include <moveit/collision_detection_bullet/collision_robot_bt.h>
+#include <tesseract_ros/ros_tesseract_utils.h>
+
+namespace collision_detection
+{
+CollisionRobotBt::CollisionRobotBt(const robot_model::RobotModelConstPtr& model, double padding, double scale)
+  : CollisionRobot(model, padding, scale)
+{
+  for (const auto& link : robot_model_->getURDF()->links_)
+  {
+    if (link.second->collision_array.size() > 0)
+    {
+      const std::vector<urdf::CollisionSharedPtr>& col_array =
+          link.second->collision_array.empty() ? std::vector<urdf::CollisionSharedPtr>(1, link.second->collision) :
+                                                 link.second->collision_array;
+
+      std::vector<shapes::ShapeConstPtr> shapes;
+      tesseract::VectorIsometry3d shape_poses;
+      tesseract::CollisionObjectTypeVector collision_object_types;
+
+      for (std::size_t i = 0; i < col_array.size(); ++i)
+      {
+        if (col_array[i] && col_array[i]->geometry)
+        {
+          shapes::ShapeConstPtr s = tesseract::tesseract_ros::constructShape(col_array[i]->geometry.get());
+          if (s)
+          {
+            shapes.push_back(s);
+            shape_poses.push_back(tesseract::tesseract_ros::urdfPose2Eigen(col_array[i]->origin));
+
+            if (s->type == shapes::MESH)
+              collision_object_types.push_back(tesseract::CollisionObjectType::ConvexHull);
+            else
+              collision_object_types.push_back(tesseract::CollisionObjectType::UseShapeType);
+          }
+        }
+      }
+      bt_manager_.addCollisionObject(link.second->name, BodyType::ROBOT_LINK, shapes, shape_poses,
+                                     collision_object_types, true);
+    }
+  }
+}
+
+CollisionRobotBt::CollisionRobotBt(const CollisionRobotBt& other) : CollisionRobot(other)
+{
+  const CollisionRobotBt& other_bt = dynamic_cast<const CollisionRobotBt&>(other);
```

Comment:
Functions like this seem to make it difficult to remove tesseract dependencies later without copying a lot of code.

Reply:
That's been the plan. Much of what Tesseract does (the contact managers) will have to be included, either by copying/porting code, or by dependencies. Rather than have Jens port every piece individually right now, we decided to keep much of it dependencies while Jens is running benchmarks to see what interface changes we might want to make as well.